// @deck.gl/test-utils/dist/esm/snapshot-test-runner.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _get from "@babel/runtime/helpers/esm/get";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import TestRunner from './test-runner';
import { getBoundingBoxInPage } from './utils/dom';
var DEFAULT_TEST_OPTIONS = {
  imageDiffOptions: {}
};
var DEFAULT_TEST_CASE = {
  name: 'Unnamed snapshot test',
  props: {},
  onAfterRender: function onAfterRender(_ref) {
    var deck = _ref.deck,
        layers = _ref.layers,
        done = _ref.done;

    if (layers.every(function (layer) {
      return layer.isLoaded;
    })) {
      done();
    }
  },
  goldenImage: ''
};

var SnapshotTestRunner = function (_TestRunner) {
  _inherits(SnapshotTestRunner, _TestRunner);

  var _super = _createSuper(SnapshotTestRunner);

  function SnapshotTestRunner(props) {
    var _this;

    _classCallCheck(this, SnapshotTestRunner);

    _this = _super.call(this, props);
    _this.isDiffing = false;
    Object.assign(_this.testOptions, DEFAULT_TEST_OPTIONS);
    return _this;
  }

  _createClass(SnapshotTestRunner, [{
    key: "initTestCase",
    value: function initTestCase(testCase) {
      _get(_getPrototypeOf(SnapshotTestRunner.prototype), "initTestCase", this).call(this, testCase);

      if (!testCase.goldenImage) {
        throw new Error("Test case ".concat(testCase.name, " does not have golden image"));
      }
    }
  }, {
    key: "runTestCase",
    value: function runTestCase(testCase, onDone) {
      var deck = this.deck;
      deck.setProps(Object.assign({}, this.props, testCase, {
        onAfterRender: function onAfterRender() {
          testCase.onAfterRender({
            deck: deck,
            layers: deck.props.layers,
            done: onDone
          });
        }
      }));
    }
  }, {
    key: "shouldRender",
    value: function shouldRender() {
      return !this.isDiffing;
    }
  }, {
    key: "assert",
    value: function assert(testCase) {
      var _this2 = this;

      if (this.isDiffing) {
        return;
      }

      this.isDiffing = true;
      var diffOptions = Object.assign({}, this.testOptions.imageDiffOptions, testCase.imageDiffOptions, {
        goldenImage: testCase.goldenImage,
        region: getBoundingBoxInPage(this.deck.canvas)
      });
      window.browserTestDriver_captureAndDiffScreen(diffOptions).then(function (result) {
        if (result.success) {
          _this2._pass(result);
        } else {
          _this2._fail(result);
        }

        _this2.isDiffing = false;

        _this2._next();
      });
    }
  }, {
    key: "defaultTestCase",
    get: function get() {
      return DEFAULT_TEST_CASE;
    }
  }]);

  return SnapshotTestRunner;
}(TestRunner);

export { SnapshotTestRunner as default };
