// @deck.gl/test-utils/dist/es6/snapshot-test-runner.js
import TestRunner from './test-runner';
import { getBoundingBoxInPage } from './utils/dom';
const DEFAULT_TEST_OPTIONS = {
  imageDiffOptions: {}
};
const DEFAULT_TEST_CASE = {
  name: 'Unnamed snapshot test',
  props: {},
  onAfterRender: ({
    deck,
    layers,
    done
  }) => {
    if (layers.every(layer => layer.isLoaded)) {
      done();
    }
  },
  goldenImage: ''
};
export default class SnapshotTestRunner extends TestRunner {
  constructor(props) {
    super(props);
    this.isDiffing = false;
    Object.assign(this.testOptions, DEFAULT_TEST_OPTIONS);
  }

  get defaultTestCase() {
    return DEFAULT_TEST_CASE;
  }

  initTestCase(testCase) {
    super.initTestCase(testCase);

    if (!testCase.goldenImage) {
      throw new Error("Test case ".concat(testCase.name, " does not have golden image"));
    }
  }

  runTestCase(testCase, onDone) {
    const {
      deck
    } = this;
    deck.setProps(Object.assign({}, this.props, testCase, {
      onAfterRender: () => {
        testCase.onAfterRender({
          deck,
          layers: deck.props.layers,
          done: onDone
        });
      }
    }));
  }

  shouldRender() {
    return !this.isDiffing;
  }

  assert(testCase) {
    if (this.isDiffing) {
      return;
    }

    this.isDiffing = true;
    const diffOptions = Object.assign({}, this.testOptions.imageDiffOptions, testCase.imageDiffOptions, {
      goldenImage: testCase.goldenImage,
      region: getBoundingBoxInPage(this.deck.canvas)
    });
    window.browserTestDriver_captureAndDiffScreen(diffOptions).then(result => {
      if (result.success) {
        this._pass(result);
      } else {
        this._fail(result);
      }

      this.isDiffing = false;

      this._next();
    });
  }

}
